### PR TITLE
*: fix `TestConcurrentlyReset`

### LINKED ISSUE #6275: `TestMicroserviceTSOServer` failed
### Which jobs are failing
```
--- FAIL: TestMicroserviceTSOServer (16.05s)
    --- FAIL: TestMicroserviceTSOServer/TestConcurrentlyReset (0.00s)
        server_test.go:116: 
            	Error Trace:	/home/runner/work/pd/pd/tests/integrations/tso/server_test.go:116
            	            				/home/runner/work/pd/pd/tests/integrations/tso/server_test.go:159
            	            				/opt/hostedtoolcache/go/1.20.1/x64/src/runtime/asm_amd64.s:1598
            	Error:      	Error "[PD:tso:ErrResetUserTimestamp]reset user timestamp failed, the specified ts is too larger than now" does not contain "is smaller than now"
            	Test:       	TestMicroserviceTSOServer/TestConcurrentlyReset
        server_test.go:116: 
            	Error Trace:	/home/runner/work/pd/pd/tests/integrations/tso/server_test.go:116
            	            				/home/runner/work/pd/pd/tests/integrations/tso/server_test.go:159
            	            				/opt/hostedtoolcache/go/1.20.1/x64/src/runtime/asm_amd64.s:1598
            	Error:      	Error "[PD:tso:ErrResetUserTimestamp]reset user timestamp failed, the specified ts is too larger than now" does not contain "is smaller than now"
            	Test:       	TestMicroserviceTSOServer/TestConcurrentlyReset
FAIL
```

### CI link
https://github.com/tikv/pd/actions/runs/4625860016/jobs/8182005937?pr=6268

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: Close #6275.

### What is changed and how does it work?

There might be a small time window that TSO in memory has been reset but the request still pass the leadership check.

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

**Review by JmPotato — APPROVED** [review]

**Review by lhy1024 — APPROVED** [review]

File: pkg/tso/global_allocator.go [hunk 1]
```diff
@@ -543,6 +543,11 @@ func (gta *GlobalTSOAllocator) campaignLeader() {
 	}()
 
 	gta.member.EnableLeader()
+	defer resetLeaderOnce.Do(func() {
```
Comment (binshi-bing) [inline pkg/tso/global_allocator.go:546]: This seems to bring no difference. You added defer resetLeaderOnce.Do(... at the bottom of the code shown below, but it's already invoked so, if my understanding is corrected, what you added is duplicated and brings no difference. This piece of code refers to pd server.campaignLeader() in which "defer resetLeaderOnce.Do(..." is invoked twice, and I thought the second one is duplicate so I intentionally didn't invoke it after line 545 gta.member.EnableLeader(). Do I have misunderstanding? Maybe there is something that I should know but I don't know.

```
	ctx, cancel := context.WithCancel(gta.ctx)
	var resetLeaderOnce sync.Once
	defer resetLeaderOnce.Do(func() {
		cancel()
		gta.member.ResetLeader()
	})

	// maintain the the leadership, after this, TSO can be service.
	gta.member.KeepLeader(ctx)
        ...

	gta.member.EnableLeader()
	defer resetLeaderOnce.Do(func() {
		cancel()
		gta.member.ResetLeader()
	})
```

**Review by binshi-bing — COMMENTED** [review]
left one comment. @rleungx @JmPotato

File: pkg/tso/global_allocator.go (hunk 1 above)
Comment (rleungx) [reply pkg/tso/global_allocator.go:546]: As mentioned in the PR description, we need to make sure that ResetLeader happens before ResetAllocatorGroup.

**Pull request merged**